landoapi/hgexports.py

```diff
@@ -163,6 +163,129 @@ def write_commit_description(self, f: io.BytesIO):
         """Writes the commit description to the specified file object."""
         f.write(self.commit_description())
 
+    def get_diff(self) -> bytes:
+        """Return the diff for this patch."""
+        try:
+            diff = []
+            line_no = 0
+            for line in self.patch:
+                line_no += 1
+
+                if self.diff_start_line:
+                    if line_no == self.diff_start_line:
+                        diff.append(line)
+                        break
+                else:
+                    if self._is_diff_line(line):
+                        diff.append(line)
+                        break
+
+            while 1:
+                buf = self.patch.read(16 * 1024)
+                if not buf:
+                    break
+                diff.append(buf)
+
+            return b"".join(diff)
+        finally:
+            self.patch.seek(0)
+
+    def write_diff(self, file_obj: io.BytesIO):
+        """Writes the diff to the specified file object."""
+        file_obj.write(self.get_diff())
+
+
+class GitPatchHelper(object):
+    """Helper class for parsing Mercurial patches/exports."""
+
+    def __init__(self, fileobj: io.BytesIO):
```

Comment:
`__init__` is a little too large -- it could be split into different instance methods based on functionality. It would be better if it was easy to figure out what `__init__` is doing by glancing at a few calls to helper methods instead.

Reply:
The intention of my comment was to encourage breaking down the functionality into separate methods, not moving it all to a single method -- since that does not alleviate the fact that there are a few too many things happening in `parse_patch` that could be broken down into different methods, and the readability (and code flow) should be improved at that point.

Reply:
As discussed during our team meeting, while updating this patch to address these comments I realized my existing solution of looking for specific headers in a specific order was not robust enough, as the user's underlying config files may add more headers to the produced patch than I was seeing in local testing.

To fix this I have changed the initial parsing of the patch to use the `email` package, as `git format-patch` patches are formatted in Unix mbox format. They are essentially email messages. So the updated parsing uses the `EmailMessage` object to get the values of the appropriate headers, and we parse the email body content for extended commit description and the contents of the diff.

I updated the `PatchHelper` class and subclasses to use `str` instead of `bytes` everywhere. I also attempted to fix this inconsistency in other parts of the code (ie make places like `HgRepo.apply_patch` no longer take an `io.BytesIO`) but the problem is a bit deeper than that and I decided it is out of scope. Despite using `str` in most places in the code, the patch itself is `LargeBinary` in the database, so we are still expecting `str` in some areas and `bytes` in others. I will file a bug to track the work of fleshing that out and being more consistent with the types we use to represent patches in the future.